File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-1.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode* seekEnd = head;
        ListNode* nthNode = head;
        ListNode dummy;
        dummy.next = head;

        while (n > 0) {
            seekEnd = seekEnd->next;
            n--;
        }
        ListNode* prev = &dummy;
        while (seekEnd) {
            seekEnd = seekEnd->next;
            prev = nthNode;
            nthNode = nthNode->next;
        }
        prev->next = nthNode->next;
        return dummy.next;
    }
};

File: Data Structures & Algorithms/reorder-linked-list/submission-1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        seekEnd = head
        middle = head
        while seekEnd.next and seekEnd.next.next:
            seekEnd = seekEnd.next.next
            middle = middle.next
        rHalf = middle.next
        middle.next = None
        prevNode = None
        while rHalf:
            nextNode = rHalf.next
            rHalf.next = prevNode
            prevNode = rHalf
            rHalf = nextNode
        rHalfReversed = prevNode
        while rHalfReversed:
            rNode = rHalfReversed
            rHalfReversed = rHalfReversed.next
            rNode.next = head.next
            head.next = rNode
            head = rNode.next
            rNode.next = head
